framework/src/media/media_recorder.c

```diff
@@ -0,0 +1,342 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+
+#include <media/media_recorder.h>
+#include <media/media_utils.h>
+#include <pthread.h>
+#include <unistd.h>
+#include <stdio.h>
+#include <fcntl.h>
+
+#define IS_SOCKET(fd) (fd > CONFIG_NFILE_DESCRIPTORS)
+#define RECORD_LOCK() do { pthread_mutex_lock(&g_rc.mutex_record); } while (false)
+#define RECORD_UNLOCK() do { pthread_mutex_unlock(&g_rc.mutex_record); } while (false)
+
+struct media_record_context_s {
+	pthread_t           pth;
+	bool                thread_running;
+	pthread_mutexattr_t mutex_attr_record;
+	pthread_mutex_t     mutex_record;
+	sem_t               sem_pause;
+	struct pcm         *pcmin;
+	struct pcm_config  *pcmconf;
+	unsigned int        buffer_size;
+	int                 fd;
+	media_format_e      format;
+	record_state_t      state;
+};
+
+static struct media_record_context_s g_rc;
+static bool b_started = false;
+
+int threadfunc_media_recording(void *args);
+
+/**
+ * Public Functions
+ */
+record_result_t media_record(int fd)
+{
+	pthread_attr_t attr;
+	struct sched_param sparam;
+
+	g_rc.fd = fd;
+
+	if (IS_SOCKET(g_rc.fd)) {
+		char instruction = 0x04;
+		send(g_rc.fd, &instruction, 1, 0);	
+	}
+	
+	RECORD_LOCK();
+
+	if (g_rc.state != RECORD_STOPPING && g_rc.state != RECORD_AVAILABLE) {
+		RECORD_UNLOCK();
+		return RECORD_ALREADY_INITIALIZED;
+	}
+
+	if (g_rc.state == RECORD_STOPPING) {
+		pthread_join(g_rc.pth, NULL);
+		g_rc.state = RECORD_AVAILABLE;
+	}
+
+	if (g_rc.state == RECORD_AVAILABLE) {
+		g_rc.fd = fd;
+		g_rc.thread_running = true;
+
+		if (pthread_attr_init(&attr) != 0) {
+			goto error_thread_init;
+		}
+
+		sparam.sched_priority = CONFIG_MEDIA_PRIORITY;
+		if (pthread_attr_setschedparam(&attr, &sparam) != 0) {
+			goto error_thread_init;
+		}
+
+		if (pthread_attr_setschedpolicy(&attr, SCHED_RR) != 0) {
+			goto error_thread_init;
+		}
+
+		if (pthread_attr_setstacksize(&attr, CONFIG_MEDIA_STACK_SIZE) != 0) {
+			goto error_thread_init;
+		}
+		if (pthread_create(&g_rc.pth, &attr, (pthread_startroutine_t)threadfunc_media_recording, NULL) != 0) {
+			goto error_thread_init;
+		}
+
+		g_rc.state = RECORD_RUNNING;
+	}
+
+	RECORD_UNLOCK();
+	return RECORD_OK;
+
+error_thread_init:
+	g_rc.thread_running = false;
+	RECORD_UNLOCK();
+	return RECORD_ERROR_THREAD_INIT;
+}
+
+record_result_t media_record_set_config(int channel, int sample_rate, int pcm_format, media_format_e format)
```

Comment:
If user use specific header for stream, we can define here as a param. please consider it.
media_record_set_config(int, int, int, media_user_header_t *header)

Reply:
yes, it is possible scenario.
we will discuss about user_header and receive callback function.